File: backend/src/routes/timetable.js
const router = require('express').Router();
const supabase = require('../config/supabase');
const requireAuth = require('../middleware/auth');

router.use(requireAuth);

// GET /timetable?class_id=...&teacher_id=...
router.get('/', async (req, res) => {
  const { class_id, teacher_id } = req.query;
  let q = supabase
    .from('timetable')
    .select('id, class_id, teacher_id, day, period, subject, classes(name, class_level), teachers(name)')
    .order('day').order('period');
  if (class_id)   q = q.eq('class_id', class_id);
  if (teacher_id) q = q.eq('teacher_id', teacher_id);
  const { data, error } = await q;
  if (error) return res.status(500).json({ error: error.message });
  res.json(data || []);
});

// GET /timetable/classes — plain list, no teacher join (avoids FK ambiguity)
router.get('/classes', async (_req, res) => {
  const { data, error } = await supabase
    .from('classes')
    .select('id, name, class_level, section, display_order, class_teacher_id')
    .order('display_order');
  if (error) return res.status(500).json({ error: error.message });
  res.json(data || []);
});

// PUT /master-cell — edit all Mon–Sat slots for one class + period row
router.put('/master-cell', async (req, res) => {
  const { class_id, period, segments } = req.body || {};
  const periodNum = Number(period);
  if (!class_id || !periodNum || periodNum < 1 || periodNum > 8) {
    return res.status(400).json({ error: 'class_id and period (1–8) required' });
  }
  if (!Array.isArray(segments) || segments.length === 0) {
    return res.status(400).json({ error: 'segments array required' });
  }

  const covered = new Set();
  for (const s of segments) {
    const start = Number(s.dayStart);
    const end = Number(s.dayEnd);
    if (!s.teacher_id || !s.subject || !start || !end || start > end || start < 1 || end > 6) {
      return res.status(400).json({ error: 'Each segment needs dayStart, dayEnd (1–6), subject, teacher_id' });
    }
    for (let d = start; d <= end; d++) {
      if (covered.has(d)) return res.status(400).json({ error: `Overlapping day ${d} in segments` });
      covered.add(d);
    }
  }
  for (let d = 1; d <= 6; d++) {
    if (!covered.has(d)) {
      return res.status(400).json({ error: `Day ${d} not covered — use lines like 1-3 Maths (Name) and 4-6 English (Name)` });
    }
  }

  try {
    const { data: existing, error: fetchErr } = await supabase
      .from('timetable')
      .select('id, day, teacher_id')
      .eq('class_id', class_id)
      .eq('period', periodNum);
    if (fetchErr) throw new Error(fetchErr.message);

    const byDay = Object.fromEntries((existing || []).map((r) => [r.day, r]));

    for (let d = 1; d <= 6; d++) {
      const seg = segments.find((s) => d >= Number(s.dayStart) && d <= Number(s.dayEnd));
      if (!seg) continue;

      const { data: clash } = await supabase
        .from('timetable')
        .select('id, classes(name)')
        .eq('teacher_id', seg.teacher_id)
        .eq('day', d)
        .eq('period', periodNum)
        .neq('class_id', class_id)
        .maybeSingle();

      if (clash) {
        const cn = clash.classes?.name || 'another class';
        const { data: t } = await supabase.from('teachers').select('name').eq('id', seg.teacher_id).single();
        return res.status(400).json({
          error: `${t?.name || 'Teacher'} is already teaching ${cn} on day ${d} period ${periodNum}`,
        });
      }

      const row = byDay[d];
      if (row) {
        const { error } = await supabase
          .from('timetable')
          .update({ teacher_id: seg.teacher_id, subject: seg.subject })
          .eq('id', row.id);
        if (error) throw new Error(error.message);
      } else {
        const { error } = await supabase
          .from('timetable')
          .insert({ class_id, day: d, period: periodNum, teacher_id: seg.teacher_id, subject: seg.subject });
        if (error) throw new Error(error.message);
      }
    }

    const { data: updated, error: outErr } = await supabase
      .from('timetable')
      .select('id, class_id, teacher_id, day, period, subject, classes(name, class_level), teachers(name)')
      .eq('class_id', class_id)
      .eq('period', periodNum)
      .order('day');
    if (outErr) throw new Error(outErr.message);

    res.json({ success: true, rows: updated });
  } catch (e) {
    res.status(400).json({ error: e.message });
  }
});

// PUT — edit a single timetable cell
router.put('/:id', async (req, res) => {
  const { teacher_id, subject } = req.body;
  const { data, error } = await supabase
    .from('timetable')
    .update({ teacher_id, subject })
    .eq('id', req.params.id)
    .select('*, teachers(name)')
    .single();
  if (error) return res.status(400).json({ error: error.message });
  res.json(data);
});

// DELETE all rows (called before re-applying a CP run)
router.delete('/all', async (_req, res) => {
  const { error } = await supabase
    .from('timetable')
    .delete()
    .neq('id', '00000000-0000-0000-0000-000000000000');
  if (error) return res.status(400).json({ error: error.message });
  res.json({ success: true });
});

module.exports = router;

File: backend/app/api/routers/timetable.py
"""Timetable — read grid, master-cell editing, single-cell edit, delete all."""
from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse

from app.db import database as db
from app.core.security import require_auth

router = APIRouter(dependencies=[Depends(require_auth)])

ZERO_UUID = "00000000-0000-0000-0000-000000000000"


def _shape_row(r):
    """Match the Supabase nested-join shape used by the frontend."""
    return {
        "id": str(r["id"]),
        "class_id": str(r["class_id"]) if r.get("class_id") else None,
        "teacher_id": str(r["teacher_id"]) if r.get("teacher_id") else None,
        "day": r["day"],
        "period": r["period"],
        "subject": r["subject"],
        "classes": {"name": r.get("class_name"), "class_level": r.get("class_level")}
        if r.get("class_name") is not None else None,
        "teachers": {"name": r.get("teacher_name")} if r.get("teacher_name") is not None else None,
    }


@router.get("/classes")
async def timetable_classes():
    rows = db.query(
        """SELECT id, name, class_level, section, display_order, class_teacher_id
           FROM classes ORDER BY display_order"""
    )
    return rows


@router.get("/")
async def get_timetable(class_id: str | None = None, teacher_id: str | None = None):
    where, params = [], []
    if class_id:
        where.append("t.class_id = %s")
        params.append(class_id)
    if teacher_id:
        where.append("t.teacher_id = %s")
        params.append(teacher_id)
    clause = ("WHERE " + " AND ".join(where)) if where else ""
    rows = db.query(
        f"""SELECT t.id, t.class_id, t.teacher_id, t.day, t.period, t.subject,
                   c.name AS class_name, c.class_level, te.name AS teacher_name
            FROM timetable t
            LEFT JOIN classes c ON c.id = t.class_id
            LEFT JOIN teachers te ON te.id = t.teacher_id
            {clause}
            ORDER BY t.day, t.period""",
        params,
    )
    return [_shape_row(r) for r in rows]


@router.put("/master-cell")
async def master_cell(request: Request):
    body = await request.json() or {}
    class_id = body.get("class_id")
    period_num = body.get("period")
    try:
        period_num = int(period_num)
    except (TypeError, ValueError):
        period_num = 0
    segments = body.get("segments")

    if not class_id or not period_num or period_num < 1 or period_num > 8:
        return JSONResponse(status_code=400, content={"error": "class_id and period (1–8) required"})
    if not isinstance(segments, list) or len(segments) == 0:
        return JSONResponse(status_code=400, content={"error": "segments array required"})

    covered = set()
    for s in segments:
        try:
            start, end = int(s.get("dayStart")), int(s.get("dayEnd"))
        except (TypeError, ValueError):
            start, end = 0, 0
        if not s.get("teacher_id") or not s.get("subject") or not start or not end \
                or start > end or start < 1 or end > 6:
            return JSONResponse(status_code=400, content={
                "error": "Each segment needs dayStart, dayEnd (1–6), subject, teacher_id"})
        for d in range(start, end + 1):
            if d in covered:
                return JSONResponse(status_code=400, content={"error": f"Overlapping day {d} in segments"})
            covered.add(d)
    for d in range(1, 7):
        if d not in covered:
            return JSONResponse(status_code=400, content={
                "error": f"Day {d} not covered — use lines like 1-3 Maths (Name) and 4-6 English (Name)"})

    try:
        existing = db.query(
            "SELECT id, day, teacher_id FROM timetable WHERE class_id = %s AND period = %s",
            [class_id, period_num],
        )
        by_day = {r["day"]: r for r in existing}

        for d in range(1, 7):
            seg = next((s for s in segments if int(s["dayStart"]) <= d <= int(s["dayEnd"])), None)
            if not seg:
                continue

            clash = db.query_one(
                """SELECT t.id, c.name AS class_name
                   FROM timetable t LEFT JOIN classes c ON c.id = t.class_id
                   WHERE t.teacher_id = %s AND t.day = %s AND t.period = %s AND t.class_id <> %s
                   LIMIT 1""",
                [seg["teacher_id"], d, period_num, class_id],
            )
            if clash:
                cn = clash.get("class_name") or "another class"
                tname = db.query_one("SELECT name FROM teachers WHERE id = %s", [seg["teacher_id"]])
                return JSONResponse(status_code=400, content={
                    "error": f"{(tname or {}).get('name') or 'Teacher'} is already teaching {cn} "
                             f"on day {d} period {period_num}"})

            row = by_day.get(d)
            if row:
                db.execute(
                    "UPDATE timetable SET teacher_id = %s, subject = %s WHERE id = %s",
                    [seg["teacher_id"], seg["subject"], row["id"]],
                )
            else:
                db.execute(
                    """INSERT INTO timetable (class_id, day, period, teacher_id, subject)
                       VALUES (%s, %s, %s, %s, %s)""",
                    [class_id, d, period_num, seg["teacher_id"], seg["subject"]],
                )

        updated = db.query(
            """SELECT t.id, t.class_id, t.teacher_id, t.day, t.period, t.subject,
                      c.name AS class_name, c.class_level, te.name AS teacher_name
               FROM timetable t
               LEFT JOIN classes c ON c.id = t.class_id
               LEFT JOIN teachers te ON te.id = t.teacher_id
               WHERE t.class_id = %s AND t.period = %s
               ORDER BY t.day""",
            [class_id, period_num],
        )
        return {"success": True, "rows": [_shape_row(r) for r in updated]}
    except Exception as e:
        return JSONResponse(status_code=400, content={"error": str(e)})


@router.delete("/all")
async def delete_all():
    try:
        db.execute("DELETE FROM timetable WHERE id <> %s", [ZERO_UUID])
    except Exception as e:
        return JSONResponse(status_code=400, content={"error": str(e)})
    return {"success": True}


@router.put("/{row_id}")
async def update_cell(row_id: str, request: Request):
    body = await request.json()
    try:
        row = db.execute_returning(
            "UPDATE timetable SET teacher_id = %s, subject = %s WHERE id = %s RETURNING *",
            [body.get("teacher_id"), body.get("subject"), row_id],
        )
    except Exception as e:
        return JSONResponse(status_code=400, content={"error": str(e)})
    if not row:
        return JSONResponse(status_code=400, content={"error": "Row not found"})
    teacher = None
    if row.get("teacher_id"):
        teacher = db.query_one("SELECT name FROM teachers WHERE id = %s", [row["teacher_id"]])
    out = {
        "id": str(row["id"]),
        "class_id": str(row["class_id"]) if row.get("class_id") else None,
        "teacher_id": str(row["teacher_id"]) if row.get("teacher_id") else None,
        "day": row["day"],
        "period": row["period"],
        "subject": row["subject"],
        "teachers": {"name": teacher["name"]} if teacher else None,
    }
    return out
